lib/DxilDia/DxilDiaDataSource.cpp

```diff
@@ -65,11 +66,25 @@ std::unique_ptr<llvm::MemoryBuffer> getMemBufferFromStream(_In_ IStream *pStream
 }
 }  // namespace dxil_dia
 
-STDMETHODIMP dxil_dia::DataSource::loadDataFromIStream(_In_ IStream *pIStream) {
+STDMETHODIMP dxil_dia::DataSource::loadDataFromIStream(_In_ IStream *pInputIStream) {
   DxcThreadMalloc TM(m_pMalloc);
   if (m_module.get() != nullptr) {
     return E_FAIL;
   }
+
+  CComPtr<IStream> pIStream = pInputIStream;
+  CComPtr<IDxcBlob> pContainer;
+  if (SUCCEEDED(hlsl::pdb::LoadDataFromStream(m_pMalloc, pInputIStream, &pContainer))) {
+    hlsl::DxilPartHeader *PartHeader =
+      hlsl::GetDxilPartByType((hlsl::DxilContainerHeader *)pContainer->GetBufferPointer(), hlsl::DFCC_ShaderDebugInfoDXIL);
+    if (!PartHeader)
+      return E_FAIL;
+    CComPtr<IDxcBlobEncoding> pPinnedBlob;
+    IFR(hlsl::DxcCreateBlobWithEncodingFromPinned(PartHeader+1, PartHeader->PartSize, CP_UTF8, &pPinnedBlob));
+    pIStream.Release();
+    IFR(hlsl::CreateReadOnlyBlobStream(pPinnedBlob, &pIStream));
+  }
+
   m_context.reset();
   m_finder.reset();
   try {
```

Comment:
Exceptions can happen before this point as the methods above use collections that might throw them.